pkg/prometheus/agent/operator.go

```diff
@@ -0,0 +1,1563 @@
+			level.Debug(logger).Log("msg", "creating statefulset")
+			if _, err := ssetClient.Create(ctx, sset, metav1.CreateOptions{}); err != nil {
+				return errors.Wrap(err, "creating statefulset failed")
+			}
+			continue
+		}
+
+		if newSSetInputHash == existingStatefulSet.ObjectMeta.Annotations[prompkg.SSetInputHashName] {
+			level.Debug(logger).Log("msg", "new statefulset generation inputs match current, skipping any actions")
+			continue
+		}
+
+		level.Debug(logger).Log(
+			"msg", "updating current statefulset because of hash divergence",
+			"new_hash", newSSetInputHash,
+			"existing_hash", existingStatefulSet.ObjectMeta.Annotations[prompkg.SSetInputHashName],
+		)
+
+		err = k8sutil.UpdateStatefulSet(ctx, ssetClient, sset)
+		sErr, ok := err.(*apierrors.StatusError)
+
+		if ok && sErr.ErrStatus.Code == 422 && sErr.ErrStatus.Reason == metav1.StatusReasonInvalid {
+			c.metrics.StsDeleteCreateCounter().Inc()
+
+			// Gather only reason for failed update
+			failMsg := make([]string, len(sErr.ErrStatus.Details.Causes))
+			for i, cause := range sErr.ErrStatus.Details.Causes {
+				failMsg[i] = cause.Message
+			}
+
+			level.Info(logger).Log("msg", "recreating StatefulSet because the update operation wasn't possible", "reason", strings.Join(failMsg, ", "))
+
+			propagationPolicy := metav1.DeletePropagationForeground
+			if err := ssetClient.Delete(ctx, sset.GetName(), metav1.DeleteOptions{PropagationPolicy: &propagationPolicy}); err != nil {
+				return errors.Wrap(err, "failed to delete StatefulSet to avoid forbidden action")
+			}
+			continue
+		}
+
+		if err != nil {
+			return errors.Wrap(err, "updating StatefulSet failed")
+		}
+	}
+
+	ssets := map[string]struct{}{}
+	for _, ssetName := range expected {
+		ssets[ssetName] = struct{}{}
+	}
+
+	err = c.ssetInfs.ListAllByNamespace(p.Namespace, labels.SelectorFromSet(labels.Set{prompkg.PrometheusNameLabelName: p.Name}), func(obj interface{}) {
+		s := obj.(*appsv1.StatefulSet)
+
+		if _, ok := ssets[s.Name]; ok {
+			// Do not delete statefulsets that we still expect to exist. This
+			// is to cleanup StatefulSets when shards are reduced.
+			return
+		}
+
+		if c.rr.DeletionInProgress(s) {
+			return
+		}
+
+		propagationPolicy := metav1.DeletePropagationForeground
+		if err := ssetClient.Delete(ctx, s.GetName(), metav1.DeleteOptions{PropagationPolicy: &propagationPolicy}); err != nil {
+			level.Error(c.logger).Log("err", err, "name", s.GetName(), "namespace", s.GetNamespace())
+		}
+	})
+	if err != nil {
+		return errors.Wrap(err, "listing StatefulSet resources failed")
+	}
+
+	return nil
+}
+
+func (c *Operator) createOrUpdateConfigurationSecret(ctx context.Context, p *monitoringv1.PrometheusAgent, cg *prompkg.ConfigGenerator, store *assets.Store) error {
+	// If no service or pod monitor selectors are configured, the user wants to
+	// manage configuration themselves. Do create an empty Secret if it doesn't
+	// exist.
```

Comment:
We don't want to support this configuration anymore with the Agent beacause additionalScrapeConfigs fulfill the same purpose so I'd remove this special case.